[VPlan] Make canonical IV part of the region

fhahn opened the pull request:
The canonical IV is directly tied to a loop region. To directly ensure there's a single, unique canonical IV, directly define it by the region.

Depends on https://github.com/llvm/llvm-project/pull/161589 (include in the PR)

fhahn committed:
[VPlan] Make canonical IV part of the region

fhahn committed:
[VPlan] Don't reset canonical IV start value.

Instead of re-setting the start value of the canonical IV when
vectorizing the epilogue we can emit an Add VPInstruction to provide
canonical IV value, adjusted by the resume value from the main loop.

This is in preparation to make the canonical IV a VPValue defined by
loop regions. It ensures that the canonical IV always starts at 0.

fhahn committed:
[VPlan] Don't reset canonical IV start value.

Instead of re-setting the start value of the canonical IV when
vectorizing the epilogue we can emit an Add VPInstruction to provide
canonical IV value, adjusted by the resume value from the main loop.

This is in preparation to make the canonical IV a VPValue defined by
loop regions. It ensures that the canonical IV always starts at 0.

ayalz reviewed (commented):
Looks overall a good approach to me! Adding initial set of various comments, to be continued.

ayalz reviewed (commented):
Overall approach looks good to me! Adding initial set of various comments, to be continued.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp:
```diff
@@ -499,6 +494,10 @@ SmallVector<VPRegisterUsage, 8> llvm::calculateRegisterUsageForPlan(
     return TTICapture.getRegUsageForType(VectorType::get(Ty, VF));
   };
 
+  if (auto *CanIV = LoopRegion->getCanonicalIV())
```
LoopRegion must always have one (and be not replicating as consequence), removed the check, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -395,7 +395,7 @@ static bool canonicalHeaderAndLatch(VPBlockBase *HeaderVPB,
 static void createLoopRegion(VPlan &Plan, VPBlockBase *HeaderVPB) {
   auto *PreheaderVPBB = HeaderVPB->getPredecessors()[0];
-  auto *LatchVPBB = HeaderVPB->getPredecessors()[1];
+  auto *LatchVPBB = cast<VPBasicBlock>(HeaderVPB->getPredecessors()[1]);
```
Unrelated in the latest version, removed thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -440,14 +447,9 @@ static void addCanonicalIVRecipes(VPlan &Plan, VPBasicBlock *HeaderVPBB,
-  // Initially the induction increment is guaranteed to not wrap, but that may
-  // change later, e.g. when tail-folding, when the flags need to be dropped.
-  auto *CanonicalIVIncrement = Builder.createOverflowingOp(
+  auto CanonicalIVIncrement = Builder.createOverflowingOp(
```
Restored original code, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanValue.h:
```diff
@@ -189,6 +195,18 @@ class LLVM_ABI_FOR_TEST VPValue {
 };
 
+/// VPValues defined by a VPRegionBlock, like the canonical IV.
+class VPRegionValue : public VPValue {
```
Yep, updated, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanValue.h:
```diff
@@ -168,7 +172,9 @@ class LLVM_ABI_FOR_TEST VPValue {
   /// Returns true if this VPValue is a live-in, i.e. defined outside the VPlan.
-  bool isLiveIn() const { return !hasDefiningRecipe(); }
+  bool isLiveIn() const {
+    return !hasDefiningRecipe() && SubclassID != VPRegionValueSC;
```
As in also checking in hasDefiningRecipe? Added ,thanks

fhahn committed:
!fixup address latest comments, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-can-iv-dont-reset

fhahn committed:
!fixup address latest comments, thanks

fhahn committed:
Merge branch 'vplan-can-iv-dont-reset' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn committed:
!fixup update remaining tests.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn committed:
!fixup address latest comments, thansk

fhahn committed:
[VPlan] Add getCanonicalIVType (NFC).

fhahn committed:
Merge branch 'vplan-region-getcanonicalivtype' into vplan-caniv-defined-by-region-dont-reset-start-value-tmp

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value-tmp

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -868,18 +881,37 @@ void VPRegionBlock::print(raw_ostream &O, const Twine &Indent,
+                                     m_Add(m_Specific(CanIV), m_LiveIn())),
+                         m_VPValue())) &&
+             "invalid existing IV increment");
+    }
```
I there is a check below that checks if CanIV != nullptr, merged the 2 together below, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -868,18 +881,37 @@ void VPRegionBlock::print(raw_ostream &O, const Twine &Indent,
+      CanIVInc = VPBuilder(ExitingTerm)
+                     .createOverflowingOp(
+                         Instruction::Add, {CanIV, &Plan.getVFxUF()},
+                         {CanIVInfo.HasNUW, false}, CanIVInfo.DL, "index.next");
```
Done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1224,6 +1259,11 @@ VPlan *VPlan::duplicate() {
+  if (auto *LoopRegion = getVectorLoopRegion()) {
+    Old2NewVPValues[LoopRegion->getCanonicalIV()] =
+        NewPlan->getVectorLoopRegion()->getCanonicalIV();
+  }
```
Done thanks

fhahn replied:
Added assert, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3981,23 +3933,35 @@ class LLVM_ABI_FOR_TEST VPRegionBlock : public VPBlockBase {
-  /// An indicator whether this region is to generate multiple replicated
-  /// instances of output IR corresponding to its VPBlockBases.
-  bool IsReplicator;
+  /// Canonical IV of the loop region. If CanIV is nullptr, the region is a
```
Updated, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -801,7 +800,7 @@ static VPValue *optimizeEarlyExitInductionUser(VPlan &Plan,
   // Calculate the final index.
   VPValue *EndValue = Plan.getCanonicalIV();
-  auto CanonicalIVType = Plan.getCanonicalIV()->getScalarType();
+  auto CanonicalIVType = TypeInfo.inferScalarType(EndValue);
```
Done, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2744,9 +2764,7 @@ void VPlanTransforms::addExplicitVectorLength(
-      Instruction::Add, {OpVPEVL, EVLPhi},
-      {CanonicalIVIncrement->hasNoUnsignedWrap(),
-       CanonicalIVIncrement->hasNoSignedWrap()},
+      Instruction::Add, {OpVPEVL, EVLPhi}, {CanIVInfo.HasNUW, false},
```
Done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanUtils.cpp:
```diff
@@ -102,9 +102,7 @@ bool vputils::isUniformAcrossVFsAndUFs(VPValue *V) {
-  auto *CanonicalIV = R->getParent()->getPlan()->getCanonicalIV();
-  // Canonical IV chain is uniform.
-  if (V == CanonicalIV || V == CanonicalIV->getBackedgeValue())
+  if (isa<VPRegionValue>(V))
```
This should probably be handled below, as VPInstruction BinaryOp with all operands inform across VFs and UFs. Will do separately.

fhahn replied:
Not sure f there's a good way to add this assert here, but this is covered by the unit tests I think

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanValue.h:
```diff
@@ -89,7 +91,9 @@ class LLVM_ABI_FOR_TEST VPValue {
-    VPVRecipeSC /// A VPValue sub-class that is a VPRecipeBase.
+    VPVRecipeSC,     /// A VPValue sub-class that is a VPRecipeBase.
+    VPRegionValueSC, /// A VPValue sub-class defines the canonical IV of a loop
+                     /// region.
```
Updated, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value-tmp

fhahn committed:
!fixup address latest comments, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value-tmp

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value-tmp

fhahn committed:
!fixup fix after merge.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4378,14 +4352,10 @@ class VPlan {
+  /// Returns the canonical induction VPValue of the vector loop.
+  VPValue *getCanonicalIV() { return getVectorLoopRegion()->getCanonicalIV(); }
+  VPCanonicalIVInfo &getCanonicalIVInfo() {
+    return getVectorLoopRegion()->getCanonicalIVInfo();
```
Should be tightned now so the only allowed modification is clearing NUW

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn committed:
!fixup store defining region.

fhahn reviewed (commented):
Updated after landing https://github.com/llvm/llvm-project/pull/164127

ping :)

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8338,8 +8338,7 @@ VPlanPtr LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(
       // latter are added above for masking.
```
Replace "including canonical IV," above with "including blends,"

fhahn replied:
Done thanks

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9346,10 +9341,10 @@ static void preparePlanForMainVectorLoop(VPlan &MainPlan, VPlan &EpiPlan) {
-        {VectorTC,
-         MainPlan.getVectorLoopRegion()->getCanonicalIV()->getStartValue()},
-        {}, "vec.epilog.resume.val");
+        {VectorTC, MainPlan.getOrAddLiveIn(Constant::getNullValue(Ty))}, {},
```
Have a short getConstantInt() version now?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -846,18 +858,30 @@ void VPRegionBlock::print(raw_ostream &O, const Twine &Indent,
-        CanIV->getDebugLoc(), "index");
+  auto *ExitingLatch = cast<VPBasicBlock>(getExiting());
+  VPValue *CanIV = getCanonicalIV();
+  if (CanIV && CanIV->getNumUsers() > 0) {
```
Worth documenting somewhere that CanIV should continue to exist (until loop region dissolves) even w/o users, i.e., avoid dce, as its existence indicates region loopness, and provides other information such as type; plus users might conceptually be introduced late(?)

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -846,18 +858,30 @@ void VPRegionBlock::print(raw_ostream &O, const Twine &Indent,
+    auto *ScalarR =
+        VPBuilder(Header, Header->begin())
+            .createScalarPhi(
+                {Plan.getOrAddLiveIn(ConstantInt::get(CanIVInfo->getType(), 0)),
```
getConstantInt()?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3992,6 +3926,32 @@ class VPIRBasicBlock : public VPBasicBlock {
+  /// VPRegionValue for the canonical IV. The allocation is managed by
+  /// VPCanonicalIVInfo.
+  std::unique_ptr<VPRegionValue> CanIV;
+  Type *Ty = nullptr;
```
It might be good for VPRegionValue to holds its Type, and DL. Probably clearer after additional VPRegionValues are introduced.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4097,24 +4064,23 @@ class LLVM_ABI_FOR_TEST VPRegionBlock : public VPBlockBase {
+  /// Get the canonical IV increment instruction. If the exiting terminator
+  /// is a BranchOnCount with an IV increment, return it. Otherwise, create
+  /// a new IV increment and return it.
+  VPInstruction *getCanonicalIVIncrement();
```
getOrCreate?

ayalz replied:
?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp:
```diff
@@ -23,14 +23,6 @@ using namespace llvm;
-
   // If there's no canonical IV, retrieve the type from the trip count
   // expression.
   auto *TC = Plan.getTripCount();
```
Independent nit: seems easier to have an `else` than an early exit below.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp:
```diff
@@ -272,18 +264,20 @@ Type *VPTypeAnalysis::inferScalarType(const VPValue *V) {
     return CanonicalIVTy;
```
Should all VPValues w/o any underlying IR be region values?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -402,17 +402,32 @@ static void createLoopRegion(VPlan &Plan, VPBlockBase *HeaderVPB) {
+  DebugLoc DL = DebugLoc::getCompilerGenerated();
+  // Get type info and debug location from the scalar phi corresponding to the
+  // canonical IV for outermost loops.
+  if (PreheaderVPBB->getSinglePredecessor() == Plan.getEntry()) {
```
Better check instead of PreheaderVPBB is contained within a loop?
Otherwise, (nested) loop region is constructed with null canonical IV type?

fhahn replied:
Yep, updated to always get the info from the outermost loop header, as we only add canonical IV scalar phi to the outermost header.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -425,9 +440,7 @@ static void addCanonicalIVRecipes(VPlan &Plan, VPBasicBlock *HeaderVPBB,
                                   DebugLoc DL) {
   Value *StartIdx = ConstantInt::get(IdxTy, 0);
   auto *StartV = Plan.getOrAddLiveIn(StartIdx);
```
getConstantInt()?

fhahn replied:
Done separately in f26e8595c3073fafc822bfcdd3ca9dec32ab573b, thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -662,7 +673,7 @@ void VPlanTransforms::attachCheckBlock(VPlan &Plan, Value *Cond,
           .createNaryOp(
               VPInstruction::BranchOnCond, {CondVPV},
-              Plan.getVectorLoopRegion()->getCanonicalIV()->getDebugLoc());
+              Plan.getVectorLoopRegion()->getCanonicalIVInfo().getDebugLoc());
```
```suggestion
              Plan.getVectorLoopRegion()->getCanonicalIVDebugLoc()
```
avoiding having getCanonicalIVInfo() public?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2489,8 +2493,9 @@ void VPlanTransforms::addActiveLaneMask(
-      LoopRegion->getCanonicalIV()->users(), IsaPred<VPWidenCanonicalIVRecipe>);
+  auto *CanonicalIV = LoopRegion->getCanonicalIV();
+  auto *FoundWidenCanonicalIVUser =
+      find_if(CanonicalIV->users(), IsaPred<VPWidenCanonicalIVRecipe>);
```
nit: indepenent? needed?

fhahn replied:
restored original code thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -4331,8 +4335,9 @@ void VPlanTransforms::narrowInterleaveGroups(VPlan &Plan, ElementCount VF,
-  auto *Inc = cast<VPInstruction>(CanIV->getBackedgeValue());
+  Type *CanIVTy = VectorLoop->getCanonicalIVInfo().getType();
+  auto *Inc = cast<VPInstruction>(
+      VectorLoop->getExitingBasicBlock()->getTerminator()->getOperand(0));
```
Good to assert terminator is what we expect?

fhahn replied:
updated thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanUtils.cpp:
```diff
@@ -83,17 +83,19 @@ const SCEV *vputils::getSCEVExprForVPValue(const VPValue *V,
     return SE.getCouldNotCompute();
   }
 
+  if (auto *CanIV = dyn_cast<VPRegionValue>(V)) {
```
This assumes every region value is the canonical IV?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -103,20 +103,20 @@ VPValue::VPValue(const unsigned char SC, Value *UV, VPDef *Def)
 #if !defined(NDEBUG) || defined(LLVM_ENABLE_DUMP)
 void VPValue::print(raw_ostream &OS, VPSlotTracker &SlotTracker) const {
-  if (const VPRecipeBase *R = dyn_cast_or_null<VPRecipeBase>(Def))
+  if (const VPRecipeBase *R = getDefiningRecipe())
```
nit: can be applied independently

fhahn replied:
Done in 0767c640437a8ff02e431a350168a13cb9056428, thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanValue.h:
```diff
@@ -60,9 +63,13 @@ class LLVM_ABI_FOR_TEST VPValue {
+  /// or region in VPlan, i.e. is a live-in.
+  union {
+    VPDef *Def;
+    VPRegionBlock *DefiningRegion;
```
If DefiningRegion is held in the VPValue base class, so should `getDefiningRegion()`, complementing `getDefiningRecipe()`?

If `hasDefiningRegion()` is also supported, complementing `hasDefiningRecipe()`, is it worth having a VPRegionValue as a derived class? OTOH, would be consistent to also have a VPLiveIn as another derived class of VPValue?

ayalz replied:
Would be good for the new region-defined VPValues to have an API that is consistent with the API's of live-ins and recipe-defined VPValues. Aligning the implementation to hold UnderlyingValue inside the union seems like a good cleanup either before or after this patch.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -870,6 +894,24 @@ void VPRegionBlock::dissolveToCFGLoop() {
   VPBlockUtils::connectBlocks(ExitingLatch, Header);
 }
 
+VPInstruction *VPRegionBlock::getCanonicalIVIncrement() {
```
Yep, added a TODO

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -870,6 +894,24 @@ void VPRegionBlock::dissolveToCFGLoop() {
+                                     m_c_Add(m_Specific(CanIV), m_LiveIn())),
+                         m_VPValue())) &&
+           "invalid existing IV increment");
+  }
```
Yep I think it should be fine, alternative is capturing the result of the match in a variable, marking it as maybe unusued and use that in the assert

fhahn committed:
!fixup address comments, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

Conflicts:
	llvm/lib/Transforms/Vectorize/VPlan.cpp

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -103,20 +103,20 @@ VPValue::VPValue(const unsigned char SC, Value *UV, VPDef *Def)
-  if (Def)
-    Def->removeDefinedValue(this);
+  if (VPDef *D = getDefiningRecipe())
+    D->removeDefinedValue(this);
```
Done in 0767c640437a8ff02e431a350168a13cb9056428, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -103,20 +103,20 @@ VPValue::VPValue(const unsigned char SC, Value *UV, VPDef *Def)
 
 void VPValue::dump() const {
-  const VPRecipeBase *Instr = dyn_cast_or_null<VPRecipeBase>(this->Def);
+  const VPRecipeBase *Instr = getDefiningRecipe();
```
Done in 0767c640437a8ff02e431a350168a13cb9056428, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn committed:
!fixup fix formatting

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1491,9 +1544,12 @@ void VPSlotTracker::assignNames(const VPlan &Plan) {
+    if (auto *VPBB = dyn_cast<VPBasicBlock>(VPB))
+      assignNames(VPBB);
+    else if (!cast<VPRegionBlock>(VPB)->isReplicator())
+      assignName(cast<VPRegionBlock>(VPB)->getCanonicalIV());
```
Can alternatively simplify into:
```
    else if (auto *CanIV = cast<VPRegionBlock>(VPB)->getCanonicalIV())
      assignName(CanIV);
```

fhahn replied:
updated thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4170,23 +4125,32 @@ class LLVM_ABI_FOR_TEST VPRegionBlock : public VPBlockBase {
+  /// Holds the Canonical IV of the loop region along with additional
+  /// information. If CanIVInfo is nullptr, the region is a replicating region.
+  /// Loop regions retain their canonical IVs until they are dissolved, even if
+  /// they may not have any users.
```
```suggestion
  /// they have no users.
```

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4257,24 +4221,23 @@ class LLVM_ABI_FOR_TEST VPRegionBlock : public VPBlockBase {
-    return const_cast<VPRegionBlock *>(this)->getCanonicalIV();
-  }
+  /// Get the canonical IV increment instruction. If the exiting terminator
+  /// is a BranchOnCount with an IV increment, return it. Otherwise, create
```
```suggestion
  /// is a BranchOnCount with an IV increment, return this increment. Otherwise, create
```

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4257,24 +4221,23 @@ class LLVM_ABI_FOR_TEST VPRegionBlock : public VPBlockBase {
+
+  DebugLoc getCanonicalIVDebugLoc() const { return CanIVInfo->getDebugLoc(); }
+  bool hasCanonicalIVNUW() const { return CanIVInfo->hasNUW(); }
+  void clearCanonicalIVNUW() { CanIVInfo->clearNUW(); }
```
Empty lines in between?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4448,11 +4414,11 @@ class VPlan {
   /// Create a new loop region with \p Name and entry and exiting blocks set
```
?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2774,13 +2779,15 @@ void VPlanTransforms::addExplicitVectorLength(
-  VPValue *StartV = CanonicalIVPHI->getStartValue();
+  auto *CanonicalIV = LoopRegion->getCanonicalIV();
+  auto &CanIVInfo = LoopRegion->getCanonicalIVInfo();
+  auto *CanIVTy = CanIVInfo.getType();
```
?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2774,13 +2779,15 @@ void VPlanTransforms::addExplicitVectorLength(
+  auto *CanonicalIV = LoopRegion->getCanonicalIV();
+  auto &CanIVInfo = LoopRegion->getCanonicalIVInfo();
+  auto *CanIVTy = CanIVInfo.getType();
+  VPValue *StartV = Plan.getOrAddLiveIn(ConstantInt::getNullValue(CanIVTy));
```
?

fhahn replied:
updated, thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2811,15 +2829,15 @@ void VPlanTransforms::canonicalizeEVLLoops(VPlan &Plan) {
-  CanonicalIVIncrement->eraseFromParent();
-  CanonicalIV->eraseFromParent();
+  if (match(Backedge,
+            m_c_Add(m_Specific(CanonicalIV), m_Specific(&Plan.getVFxUF())))) {
```
In that case, what does `Backedge` hold?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanValue.h:
```diff
@@ -89,7 +96,9 @@ class LLVM_ABI_FOR_TEST VPValue {
                /// that defines multiple values.
-    VPVRecipeSC /// A VPValue sub-class that is a VPRecipeBase.
+    VPVRecipeSC,     /// A VPValue sub-class that is a VPRecipeBase.
+    VPRegionValueSC, /// A VPValue sub-class that is defined by a region, like
```
```suggestion
    VPVRegionSC, /// A VPValue sub-class that is defined by a region, like
```
more consistent with `VPVRecipeSC`?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanValue.h:
```diff
@@ -189,6 +203,22 @@ class LLVM_ABI_FOR_TEST VPValue {
+  /// Returns the region that defines this value.
+  VPRegionBlock *getDefiningRegion() const { return DefiningRegion; }
+
+  static inline bool classof(const VPValue *V) {
```
Passing the ID explicitly is indeed one way to resolve the confusing ambiguity in constructing (2) vs. (2.1).

Should there be a derived class to represent live-ins, say `VPIRValue`, complementing `VPRegionValue`?

Note that `isa<VPSingleDefRecipe>` applies to most recipes that define a single `VPValue`, but it excludes `VPWiden[EVL]LoadRecipe`. Should there be a `VPRecipeValue` which derives from `VPValue` only, and serves derivatives `VPSingleDefRecipe` and `VPWiden[EVL]LoadRecipe`?

This will allow checking `isa<VPIRValue, VPRegionValue, VPRecipeValue>(VPV)` consistently.

fhahn replied:
This should be taken care of, after merging VPValue refactoring.

fhahn committed:
merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn committed:
!fixup address comments, thanks

fhahn committed:
No

fhahn committed:
Step

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4097,24 +4064,23 @@ class LLVM_ABI_FOR_TEST VPRegionBlock : public VPBlockBase {
+  /// a new IV increment and return it.
+  VPInstruction *getCanonicalIVIncrement();
 
-  /// Return the type of the canonical IV for loop regions.
```
done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp:
```diff
@@ -448,12 +442,13 @@ SmallVector<VPRegisterUsage, 8> llvm::calculateRegisterUsageForPlan(
 
         // If this recipe is outside the loop then record it and continue.
-        if (!DefR) {
+        if (!DefR && !isa<VPRegionValue>(U)) {
```
Restructured checks thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1780,6 +1781,8 @@ static bool simplifyBranchConditionForVFAndUF(VPlan &Plan, ElementCount BestVF,
       HeaderR.eraseFromParent();
     }
+    VectorRegion->getCanonicalIV()->replaceAllUsesWith(Plan.getOrAddLiveIn(
+        ConstantInt::getNullValue(VectorRegion->getCanonicalIVType())));
```
doen thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -4331,8 +4335,9 @@ void VPlanTransforms::narrowInterleaveGroups(VPlan &Plan, ElementCount VF,
   // original iteration.
-  auto *CanIV = VectorLoop->getCanonicalIV();
-  auto *Inc = cast<VPInstruction>(CanIV->getBackedgeValue());
+  Type *CanIVTy = VectorLoop->getCanonicalIVInfo().getType();
```
done thanks

fhahn committed:
!fixup address comments, thanks

fhahn reviewed (commented):
Finally had time to update after merging VPValue refactoring. Comments should be addressed, thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4443,24 +4403,33 @@ class LLVM_ABI_FOR_TEST VPRegionBlock : public VPBlockBase {
+  /// Get the canonical IV increment instruction. If the exiting terminator
+  /// is a BranchOnCount with an IV increment, return this increment. Otherwise,
+  /// return nullptr.
+  VPInstruction *getOrCreateCanonicalIVIncrement();
```
```suggestion
  VPInstruction *getCanonicalIVIncrement() const;
```
this appears to only get a CanIVInc if one exists, returning null otherwise?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp:
```diff
@@ -457,13 +449,11 @@ SmallVector<VPRegisterUsage, 8> llvm::calculateRegisterUsageForPlan(
         if (!DefR) {
-          LoopInvariants.insert(U);
+          auto *IRV = dyn_cast<VPIRValue>(U);
+          // If this recipe is outside the loop then record it and continue.
```
Worth updating the comment - the "continue" is now separated from the "record".

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp:
```diff
@@ -510,6 +500,10 @@ SmallVector<VPRegisterUsage, 8> llvm::calculateRegisterUsageForPlan(
     return TTICapture.getRegUsageForType(VectorType::get(Ty, VF));
   };
 
+  VPValue *CanIV = LoopRegion->getCanonicalIV();
```
```suggestion
  VPValue *CanIV = LoopRegion->getCanonicalIV();
  // Note: canonical IV's are retained even if they have no users.
```

fhahn replied:
updated thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -428,18 +428,28 @@ static bool canonicalHeaderAndLatch(VPBlockBase *HeaderVPB,
+      Plan.getEntry()->getSuccessors()[1]->getSingleSuccessor());
+  ScalarCanIV = cast<VPPhi>(&OutermostHeaderVPBB->front());
+  CanIVTy = ScalarCanIV->getOperand(0)->getLiveInIRValue()->getType();
+  DL = ScalarCanIV->getDebugLoc();
```
```suggestion
  DebugLoc DL = ScalarCanIV->getDebugLoc();
```

fhahn replied:
Done thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -428,18 +428,28 @@ static bool canonicalHeaderAndLatch(VPBlockBase *HeaderVPB,
-  // HeaderVPB and LatchVPBB have been disconnected from their
-  // predecessors/successors.
-  auto *R = Plan.createLoopRegion();
+  // LatchExitVPB, taking care to preserve the original predecessor & successor
```
Is this change from "exit blocks" to "LatchExitVPB" related/separate?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -448,6 +458,10 @@ static void createLoopRegion(VPlan &Plan, VPBlockBase *HeaderVPB) {
+  if (Plan.getEntry() == PreheaderVPBB->getSinglePredecessor()) {
+    ScalarCanIV->replaceAllUsesWith(R->getCanonicalIV());
+    ScalarCanIV->eraseFromParent();
+  }
```
Deserves a comment.

fhahn replied:
added thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -5379,8 +5374,13 @@ void VPlanTransforms::narrowInterleaveGroups(VPlan &Plan, ElementCount VF,
+      match(VectorLoop->getExitingBasicBlock()->getTerminator(),
+            m_BranchOnCount(m_VPInstruction(Inc),
+                            m_Specific(&Plan.getVectorTripCount())));
+  assert(SupportedTerminator && "supported terminator");
```
```suggestion
  assert(SupportedTerminator && "Unsupported terminator");
```

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -870,6 +889,37 @@ void VPRegionBlock::dissolveToCFGLoop() {
+      CanIVInc = nullptr;
+  }
+  assert(!CanIVInc || match(CanIVInc, m_CanIVIncrement()) &&
+                          "invalid existing IV increment");
```
added early exits, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4338,6 +4269,26 @@ class VPIRBasicBlock : public VPBasicBlock {
+  /// VPRegionValue for the canonical IV. The allocation is managed by
+  /// VPCanonicalIVInfo.
+  std::unique_ptr<VPRegionValue> CanIV;
+  bool HasNUW = true;
```
added thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4443,24 +4403,33 @@ class LLVM_ABI_FOR_TEST VPRegionBlock : public VPBlockBase {
+  /// Return the type of the canonical IV for loop regions
+  Type *getCanonicalIVType() const { return CanIVInfo->getVPValue()->getType(); }
+
+  DebugLoc getCanonicalIVDebugLoc() const {
```
done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -428,18 +428,28 @@ static bool canonicalHeaderAndLatch(VPBlockBase *HeaderVPB,
 static void createLoopRegion(VPlan &Plan, VPBlockBase *HeaderVPB) {
+  VPPhi *ScalarCanIV = nullptr;
+  Type *CanIVTy = nullptr;
+  DebugLoc DL = DebugLoc::getCompilerGenerated();
```
moved thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -428,18 +428,28 @@ static bool canonicalHeaderAndLatch(VPBlockBase *HeaderVPB,
+  auto *OutermostHeaderVPBB = cast<VPBasicBlock>(
+      Plan.getEntry()->getSuccessors()[1]->getSingleSuccessor());
+  ScalarCanIV = cast<VPPhi>(&OutermostHeaderVPBB->front());
+  CanIVTy = ScalarCanIV->getOperand(0)->getLiveInIRValue()->getType();
```
done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2852,15 +2855,14 @@ static VPActiveLaneMaskPHIRecipe *addVPLaneMaskPhiAndUpdateExitBranch(
-      cast<VPInstruction>(CanonicalIVPHI->getBackedgeValue());
+  VPValue *CanonicalIV = TopRegion->getCanonicalIV();
+  VPValue *StartV = Plan.getConstantInt(TopRegion->getCanonicalIVType(), 0);
+  auto *CanonicalIVIncrement = TopRegion->getOrCreateCanonicalIVIncrement();
```
Now it always returns a canonical IV increment

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -3302,8 +3304,7 @@ void VPlanTransforms::addExplicitVectorLength(
   auto *NextEVLIV = Builder.createAdd(
       OpVPEVL, EVLPhi, CanonicalIVIncrement->getDebugLoc(), "index.evl.next",
-      {CanonicalIVIncrement->hasNoUnsignedWrap(),
-       CanonicalIVIncrement->hasNoSignedWrap()});
```
We never set NSW for the canonical IV

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanUtils.cpp:
```diff
@@ -148,6 +148,16 @@ const SCEV *vputils::getSCEVExprForVPValue(const VPValue *V,
+           "RegionValue must be canonical IV");
+    if (!L)
+      return SE.getCouldNotCompute();
+    const SCEV *Start = SE.getZero(RV->getDefiningRegion()->getCanonicalIVType());
```
updated thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -104,7 +104,7 @@ bool VPlanVerifier::verifyPhiRecipes(const VPBasicBlock *VPBB) {
     if (isa<VPEVLBasedIVPHIRecipe>(RecipeI) &&
-        !isa_and_nonnull<VPCanonicalIVPHIRecipe>(std::prev(RecipeI))) {
+        RecipeI->getIterator() != VPBB->begin()) {
       errs() << "EVL based IV is not immediately after canonical IV\n";
```
updated thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3728,66 +3721,6 @@ class VPExpandSCEVRecipe : public VPSingleDefRecipe {
-
-  /// Return the cost of this VPCanonicalIVPHIRecipe.
-  InstructionCost computeCost(ElementCount VF,
-                              VPCostContext &Ctx) const override {
```
agreed, although not sure where a good place may be

fhahn committed:
!fixup address latest comments, thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2229,6 +2222,9 @@ class VPVectorPointerRecipe : public VPRecipeWithIRFlags {
 ///  * VPEVLBasedIVPHIRecipe
+///
+///  Note that the canonical IV is modeled as a VPRegionValue associated with
+///  lop regions.
```
```suggestion
///  loop regions.
```

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9078,10 +9076,10 @@ static void preparePlanForMainVectorLoop(VPlan &MainPlan, VPlan &EpiPlan) {
-        {VectorTC,
-         MainPlan.getVectorLoopRegion()->getCanonicalIV()->getStartValue()},
-        {}, "vec.epilog.resume.val");
+        {VectorTC, MainPlan.getConstantInt(Ty, 0)}, {},
```
Is this change independent, i.e., start value of canonical IV of main plan is expected to be zero (can be asserted), and can be (re)materialized?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanUtils.cpp:
```diff
@@ -148,6 +148,15 @@ const SCEV *vputils::getSCEVExprForVPValue(const VPValue *V,
+    if (!L)
+      return SE.getCouldNotCompute();
+    return SE.getAddRecExpr(SE.getZero(RV->getType()), SE.getOne(RV->getType()),
+                            L, SCEV::FlagAnyWrap);
```
This retains current behavior, but could be improved it RV is known to NUW?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanUtils.h:
```diff
@@ -61,7 +61,7 @@ bool isHeaderMask(const VPValue *V, const VPlan &Plan);
 /// Checks if \p V is uniform across all VF lanes and UF parts. It is considered
 /// as such if it is either loop invariant (defined outside the vector region)
 /// or its operand is known to be uniform across all VFs and UFs (e.g.
```
```suggestion
/// or its operands are known to be uniform across all VFs and UFs (e.g.,
```
while we're here

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -3273,8 +3271,8 @@ void VPlanTransforms::addExplicitVectorLength(
-                                     {CanonicalIVIncrement->hasNoUnsignedWrap(),
-                                      CanonicalIVIncrement->hasNoSignedWrap()});
+                                     {LoopRegion->hasCanonicalIVNUW(),
+                                      /*HasNSW=*/false});
```
Yep restored the original code in the latest version

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -3323,32 +3321,29 @@ void VPlanTransforms::convertToVariableLengthStep(VPlan &Plan) {
+      R->replaceAllUsesWith(CurrentIterationIncr);
+      R->eraseFromParent();
+    }
+  }
```
yep, added new find-only helper, used here and below, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanValue.h:
```diff
@@ -163,7 +169,7 @@ class LLVM_ABI_FOR_TEST VPValue {
 
   /// Returns true if this VPValue is defined by a recipe.
-  bool hasDefiningRecipe() const { return getDefiningRecipe(); }
+  bool hasDefiningRecipe() const { return SubclassID == VPVRecipeValueSC; }
```
Now with the region value version, we can keep using getDefiningRecipe, thanks!

fhahn commented on llvm/unittests/Transforms/Vectorize/VPlanPatternMatchTest.cpp:
```diff
@@ -23,32 +23,31 @@ using VPPatternMatchTest = VPlanTestBase;
   VPlan &Plan = getPlan();
+  IntegerType *I64Ty = IntegerType::get(C, 64);
+  VPRegionBlock *VPR =
+      Plan.createLoopRegion(I64Ty, DebugLoc::getCompilerGenerated(), "");
```
removed, thanks

fhahn committed:
!fixup address latest comments, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
added thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -847,19 +857,20 @@ void VPRegionBlock::print(raw_ostream &O, const Twine &Indent,
+  auto *ExitingLatch = cast<VPBasicBlock>(getExiting());
+  auto *CanIV = cast<VPRegionValue>(getCanonicalIV());
+  if (CanIV->getNumUsers() > 0) {
+    VPlan &Plan = *getPlan();
```
Updated thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4359,6 +4297,28 @@ class VPIRBasicBlock : public VPBasicBlock {
+      : CanIV(std::make_unique<VPRegionValue>(Ty, DL, Region)), HasNUW(HasNUW) {
+  }
+
+  VPRegionValue *getVPValue() { return CanIV.get(); }
```
const version should be below

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4895,13 +4874,15 @@ class VPlan {
-  /// to \p Entry and \p Exiting respectively, if set. The returned block is
-  /// owned by the VPlan and deleted once the VPlan is destroyed.
-  VPRegionBlock *createLoopRegion(const std::string &Name = "",
+  /// Create a new loop region with a canonical IV using \p CanIVTy and DL. Use
```
```suggestion
  /// Create a new loop region with a canonical IV using \p CanIVTy and \p DL. Use
```

fhahn replied:
updated thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -751,9 +762,7 @@ void VPlanTransforms::createHeaderPhiRecipes(
-  for (VPRecipeBase &R : make_early_inc_range(HeaderVPBB->phis())) {
-    if (isa<VPCanonicalIVPHIRecipe>(&R))
-      continue;
+  for (VPRecipeBase &R : make_early_inc_range(drop_begin(HeaderVPBB->phis()))) {
```
Can be committed independently?
Can assert the first phi dropped `isa<VPCanonicalIVPHIRecipe>`, and then update the assert.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanConstruction.cpp:
```diff
@@ -1137,7 +1146,7 @@ static constexpr uint32_t CheckBypassWeights[] = {1, 127};
 static void addBypassBranch(VPlan &Plan, VPBasicBlock *CheckBlockVPBB,
                             VPValue *Cond, bool AddBranchWeights) {
-  DebugLoc DL = Plan.getVectorLoopRegion()->getCanonicalIV()->getDebugLoc();
+  DebugLoc DL = Plan.getVectorLoopRegion()->getCanonicalIVDebugLoc();
```
The existing
```  DebugLoc DL = Plan.getVectorLoopRegion()->getCanonicalIV()->getDebugLoc();```
still works too, right?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2163,9 +2164,15 @@ static bool simplifyBranchConditionForVFAndUF(VPlan &Plan, ElementCount BestVF,
+    // executes exactly once.
+    auto *CanIV = VectorRegion->getCanonicalIV();
+    CanIV->replaceAllUsesWith(
+        Plan.getConstantInt(VectorRegion->getCanonicalIVType(), 0));
```
```suggestion
        Plan.getConstantInt(CanIV->getType(), 0));
```

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -3300,14 +3305,15 @@ void VPlanTransforms::addExplicitVectorLength(
 
   // Create the CurrentIteration recipe in the vector loop.
   auto *CurrentIteration =
       new VPCurrentIterationPHIRecipe(StartV, DebugLoc::getUnknown());
```
Independent: CanIV can also offer DL

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -3349,10 +3353,10 @@ void VPlanTransforms::addExplicitVectorLength(
 
-  // Replace all uses of VPCanonicalIVPHIRecipe by
+  // Replace all uses of VPCanonicalIV by
   // VPCurrentIterationPHIRecipe except for the canonical IV increment.
```
Better realign lines.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -3438,11 +3435,12 @@ void VPlanTransforms::convertEVLExitCond(VPlan &Plan) {
     return;
 
+  VPValue *CanIVInc = vputils::findCanonicalIVIncrement(
+      LoopRegion->getCanonicalIV(), &Plan.getVFxUF());
```
assert CanIVInc ain't null?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanUtils.cpp:
```diff
@@ -648,6 +643,40 @@ vputils::getMemoryLocation(const VPRecipeBase &R) {
+    return cast<VPInstruction>(R);
+
+  // If CanIV is a region value, also try to find the increment via the
+  // terminator of the exiting block.
```
If CanIV is a region value, should't CanIVInc be an Add that uses it along with VFxUF (as documented), can the code below succeed where the above failed?

fhahn replied:
Problem is transforms that adjust the canoncal IV increment separately, like narrowInterLeaveGroups. As follow-up, we could make the increment an operand of the region and not explicitly create the increment VPInstruction, until region is dissolved.

This would help to clean up multiple parts of the changes

ayalz replied:
Well, if narrowInterLeaveGroups adjusts the canonical IV increment separately, so that it adds CanIV + UF instead of CanIV + VFxUF, we still expect to "find" it, otherwise we'll reconstruct it as CanIV + VFxUF (the "create" part of getOrCreate)? Sounds like narrowInterLeaveGroups should update the meaning of VFxUF, better called "SF" (Strip-mine-Factor), to have the value UF instead of VFxUF. That way CanIVInc is always CanIV + SF.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4521,24 +4490,34 @@ class LLVM_ABI_FOR_TEST VPRegionBlock : public VPBlockBase {
+  DebugLoc getCanonicalIVDebugLoc() const {
+    return CanIVInfo->getVPValue()->getDebugLoc();
   }
+
```
added thanks

fhahn committed:
!fixup address comments, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4521,24 +4490,34 @@ class LLVM_ABI_FOR_TEST VPRegionBlock : public VPBlockBase {
   }
+
+  bool hasCanonicalIVNUW() const { return CanIVInfo->hasNUW(); }
+
```
added thanks

fhahn committed:
!fixup address latest comments, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn committed:
!fixup address latest comments, thanks

Steps

Step

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanUtils.h:
```diff
@@ -141,6 +141,10 @@ template <unsigned Opcode> static VPInstruction *findUserOf(VPValue *V) {
   return cast_or_null<VPInstruction>(findUserOf(V, m_VPInstruction<Opcode>()));
 }
 
+/// Find the canonical IV increment (CanIV + VFxUF) among users of \p CanIV or
```
This has been reworked now to always look  by a simple walk of the users of CanIV

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanUtils.cpp:
```diff
@@ -648,6 +643,36 @@ vputils::getMemoryLocation(const VPRecipeBase &R) {
 }
 
+VPInstruction *vputils::findCanonicalIVIncrement(VPValue *CanIV,
+                                                 VPValue *VFxUF) {
```
Reworked to not rely on branches

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanUtils.cpp:
```diff
@@ -648,6 +643,36 @@ vputils::getMemoryLocation(const VPRecipeBase &R) {
+                                                 VPValue *VFxUF) {
+  using namespace VPlanPatternMatch;
+  if (auto *R =
+          findUserOf(CanIV, m_c_Add(m_Specific(CanIV), m_Specific(VFxUF))))
```
Now always just walk the use list of the canonical IV, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9056,15 +9048,27 @@ static SmallVector<Instruction *> preparePlanForEpilogueVectorLoop(
   VPInstruction *Add = Builder.createAdd(IV, VPV);
-  IV->replaceAllUsesWith(Add);
-  Add->setOperand(0, IV);
+  // Replace all users of the canonical IV with the offset version, except for
```
I am not sure what kind of TOOD, the latest version should fix the issue mentioned above

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2272,6 +2264,9 @@ class VPVectorPointerRecipe : public VPRecipeWithIRFlags {
 ///  * VPEVLBasedIVPHIRecipe
+///
+///  Note that the canonical IV is modeled as a VPRegionValue associated with
+///  loop regions.
```
Updated thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4521,23 +4491,34 @@ class LLVM_ABI_FOR_TEST VPRegionBlock : public VPBlockBase {
+  /// Unsets NUW for the canonical IV increment \p Increment, for loop regions.
+  void clearCanonicalIVNUW(VPInstruction *Increment) {
+    Increment->dropPoisonGeneratingFlags();
+    CanIVInfo->clearNUW();
```
Yep

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -3504,11 +3518,15 @@ void VPlanTransforms::convertEVLExitCond(VPlan &Plan) {
+      FoundIncrement &&
+      match(CanIVInc, m_Add(m_Specific(LoopRegion->getCanonicalIV()),
+                            m_Specific(&Plan.getVFxUF()))) &&
       "Expected BranchOnCond with ICmp comparing CanIV increment with vector "
```
updated thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanUtils.cpp:
```diff
@@ -648,6 +643,52 @@ vputils::getMemoryLocation(const VPRecipeBase &R) {
+    if (!UF.isMaterialized())
+      continue;
+
+    unsigned ConcreteUF = Plan.getConcreteUF();
```
Hmm, that would be an option, but may be a bit fragile and we cannot provide similar APIs for other VPSymbolicValues, as they could RAUW with another value.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanUtils.cpp:
```diff
@@ -648,6 +643,52 @@ vputils::getMemoryLocation(const VPRecipeBase &R) {
+
+  assert((!VFxUF.isMaterialized() || Increment) &&
+         "When VFxUF has been materialized, "
+         "an UF based increment must exist");
```
added thanks!

fhahn committed:
!fixup address latest comments, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9197,15 +9187,29 @@ static SmallVector<Instruction *> preparePlanForEpilogueVectorLoop(
+  auto *BranchOnCount = cast<VPInstruction>(&VectorLatch->back());
+  assert(BranchOnCount->getOpcode() == VPInstruction::BranchOnCount &&
+         "expected BranchOnCount in exiting block");
+  Builder.setInsertPoint(BranchOnCount);
```
Updated, although I left the RAUW for both IV and increment, to handle both cases the same way.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1664,14 +1664,34 @@ static void simplifyRecipe(VPSingleDefRecipe *Def, VPTypeAnalysis &TypeInfo) {
+  // updating the users Inc and IVInc to use X and IVInc respectively.
+  if (match(Def, m_Add(m_Add(m_VPValue(X), m_VPValue()), m_VPValue(Y))) &&
+      isa<VPIRValue>(Y) && !isa<VPConstantInt>(Y) && isa<VPPhi>(X)) {
     auto *Phi = cast<VPPhi>(X);
```
Should be taken care of by pattern, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1664,14 +1664,34 @@ static void simplifyRecipe(VPSingleDefRecipe *Def, VPTypeAnalysis &TypeInfo) {
+  // Transforms this by folding the increment Y into the phi's start value and
+  // updating the users Inc and IVInc to use X and IVInc respectively.
+  if (match(Def, m_Add(m_Add(m_VPValue(X), m_VPValue()), m_VPValue(Y))) &&
+      isa<VPIRValue>(Y) && !isa<VPConstantInt>(Y) && isa<VPPhi>(X)) {
```
Yes, it may be materialized VFxUF (for scalable vectors)

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1664,14 +1664,34 @@ static void simplifyRecipe(VPSingleDefRecipe *Def, VPTypeAnalysis &TypeInfo) {
-      Def->replaceAllUsesWith(Phi);
+      for (VPUser *U : to_vector(Phi->users()))
+        if (U != IVInc->getDefiningRecipe())
+          cast<VPSingleDefRecipe>(U)->replaceAllUsesWith(Phi);
```
updated thanks

fhahn committed:
[VPlan] Add m_BinaryVPPhi pattern matcher and use in removeDeadRecipes (NFC).

Add m_BinaryVPPhi to match VPPhi instructions with exactly 2 operands,
binding the start value and backedge value. Use it in removeDeadRecipes
to simplify the dead phi cycle detection.

fhahn committed:
!fixup address comments, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -3375,14 +3400,15 @@ void VPlanTransforms::addExplicitVectorLength(
-  VPValue *StartV = CanonicalIVPHI->getStartValue();
+  auto *CanonicalIV = LoopRegion->getCanonicalIV();
+  auto *CanIVTy = CanonicalIV->getType();
+  VPValue *StartV = Plan.getConstantInt(CanIVTy, 0);
```
done thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanUtils.cpp:
```diff
@@ -649,6 +644,54 @@ vputils::getMemoryLocation(const VPRecipeBase &R) {
+    if (!UF.isMaterialized()) {
+      if (Step == &UF)
+        SetIncrement(U);
+      continue;
```
yep

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2179,9 +2200,10 @@ static bool simplifyBranchConditionForVFAndUF(VPlan &Plan, ElementCount BestVF,
-            m_BranchOnCount(m_Add(m_VPValue(), m_Specific(&Plan.getVFxUF())),
-                            m_VPValue())) ||
+  auto m_CanIVInc = m_Add(m_VPValue(), m_Specific(&Plan.getVFxUF()));
+  if (match(Term, m_BranchOnCount(
```
added thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1664,15 +1664,29 @@ static void simplifyRecipe(VPSingleDefRecipe *Def, VPTypeAnalysis &TypeInfo) {
+  // Fold the increment Y into the phi's start value, replace Def with IVInc,
+  // and if Inc exists, replace it with X.
+  if (match(Def, m_Add(m_Add(m_VPValue(X), m_VPValue()), m_VPValue(Y))) &&
+      isa<VPIRValue>(Y) && !isa<VPConstantInt>(Y) &&
```
Still curious why Y must now be non-constant.

ayalz commented on llvm/test/Transforms/LoopVectorize/VPlan/RISCV/vplan-vp-intrinsics-reduction.ll:
```diff
@@ -19,7 +19,7 @@
 define i32 @reduction(ptr %a, i64 %n, i32 %start) {
 ; IF-EVL: VPlan 'Initial VPlan for VF={1},UF>=1'
-; IF-EVL-NOT: CURRENT-ITERATION-PHI
+; IF-EVL-NOT: EXPLICIT-VECTOR-LENGTH-BASED-IV-PHI
```
??

ayalz reviewed (approved):
Ok, this LGTM, thanks!
Adding last minor nits.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

fhahn committed:
!fixup

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-caniv-defined-by-region-dont-reset-start-value

bgra8 commented:
@fhahn we (at google) have bisected a compiler assertion failure to this revision.

To reproduce build `clang` with assertions enabled at this revision and compile the following `reduced.ll` (compilation command at the end):

```ll
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

define ptr @barney(i64 %arg, ptr %arg1, ptr %arg2, ptr %arg3) #0 {
bb:
  br label %bb4

bb4:                                              ; preds = %bb5, %bb
  %phi = phi i64 [ 0, %bb ], [ %add, %bb5 ]
  %icmp = icmp slt i64 %phi, %arg
  br i1 %icmp, label %bb5, label %bb17

bb5:                                              ; preds = %bb4
  %add = add i64 %phi, 1
  %getelementptr = getelementptr [8 x i8], ptr %arg1, i64 %add
  %load = load double, ptr %getelementptr, align 8
  %getelementptr6 = getelementptr [16 x i8], ptr %arg3, i64 %add
  %load7 = load double, ptr %getelementptr6, align 8
  %getelementptr8 = getelementptr [16 x i8], ptr %arg3, i64 %phi
  %load9 = load double, ptr %getelementptr8, align 8
  %getelementptr10 = getelementptr [16 x i8], ptr %arg2, i64 %phi
  %fsub = fsub double %load7, %load9
  %fdiv = fdiv double 0.000000e+00, %load
  %fmul = fmul double %fsub, %fdiv
  call void @spam(ptr %getelementptr10, double %fmul)
  %getelementptr11 = getelementptr i8, ptr %getelementptr6, i64 8
  %load12 = load double, ptr %getelementptr11, align 8
  %getelementptr13 = getelementptr i8, ptr %getelementptr8, i64 8
  %load14 = load double, ptr %getelementptr13, align 8
  %fsub15 = fsub double %load12, %load14
  %fmul16 = fmul double %fsub15, %fdiv
  call void @blam(ptr %getelementptr10, double %fmul16)
  br label %bb4

bb17:                                             ; preds = %bb4
  ret ptr null
}

define void @spam(ptr %arg, double %arg1) {
bb:
  store double %arg1, ptr %arg, align 8
  ret void
}

define void @blam(ptr %arg, double %arg1) {
bb:
  %getelementptr = getelementptr i8, ptr %arg, i64 8
  store double %arg1, ptr %getelementptr, align 8
  ret void
}

attributes #0 = { "prefer-vector-width"="128" }
```

Compilation command:

```bash
$ clang -O2 --target=x86_64-unknown-linux-gnu -c -o /dev/null reduced.ll
```

Can you please take a look?

fhahn commented:
Thanks for the heads up, will take a look

bgra8 commented:
@fhahn any idea what's going on here? Should we revert and investigate offline?

fhahn commented:
Should already be fixed by https://github.com/llvm/llvm-project/commit/eb899dca28677ac2e72b0a85f04e9c9d04df7bbe
